### PR TITLE
[RFC] add the `?` postfix operator

### PR DESCRIPTION
**Pull request by japaric**
# DO NOT MERGE NEEDS AN APPROVED RFC

The `?` postfix operator is sugar equivalent to the `try!` macro, but is more amenable to chaining: `File::open("foo")?.metadata()?.is_dir()`. Currently the syntax is accepted in two places:
- method calls: `x.foo()?`, and
- function calls: `foo()?`

But it could be extended to idents: `let x = y?;` or `foo(x?, y?)`.

---

cc @aturon @nikomatsakis

### CODE REVIEW

**Comment by eddyb** [comment]
Is it not a properly defined as a postfix operator?
I guess it binds tightly and only identifiers and function/method calls could result in a `Result

File: src/libsyntax/ast.rs [hunk 1]
```diff
@@ -774,7 +774,10 @@ pub enum Expr_ {
     ExprRepeat(P<Expr> /* element */, P<Expr> /* count */),
 
     /// No-op: used solely so we can pretty-print faithfully
-    ExprParen(P<Expr>)
+    ExprParen(P<Expr>),
+
+    /// `expr?`
+    ExprQuestion(P<Expr>),
```
Comment (eddyb) [inline src/libsyntax/ast.rs]: Arguably this could have a better name.

**Comment by nikomatsakis** [comment]
:+1: @japaric you are on a roll here.

**Comment by aturon** [comment]
@japaric thanks so much for taking the initiative here! As you point out, we need to have some accepted RFC for this, so @nikomatsakis and I will also work with @glaebhoerl to revise the existing RFC so we can move forward.

File: src/libsyntax/ast.rs (hunk 1 above)
Comment (liigo) [inline src/libsyntax/ast.rs]: ExprTry?

Comment (glaebhoerl) [inline src/libsyntax/ast.rs]: For what's it's worth I think of `?` as the exception propagation operator.

**Comment by nikomatsakis** [comment]
I'm going to close this PR, not because I don't want it, but because the lack of RFC is still blocking us. Hopefully we'll see motion there soon now that 1.0 is out the door! Thanks @japaric for the PR, in any case.

**Pull request closed without merging**